### ISSUE TITLE
Specifying evaluation loss for early stopping

### ISSUE THREAD

**Issue by t-shoemaker** [center-for-humans-and-machines/transformer-heads#15]
First of all, thanks very much for this package. It's extremely useful.

I'd like to implement early stopping on a set of simple classification probes, e.g.:

```py
checkpoint = "gpt2"
params = get_model_params(checkpoint)

heads = (0, 6, 11)
head_configs = [
    HeadConfig(
        name=f"cls_head-{h}",
        layer_hook=h,
        in_size=params["hidden_size"],
        output_activation="linear",
        pred_for_sequence=True,
        loss_fct="cross_entropy",
        num_outputs=2
    ) for h in heads
]

model = load_headed(
    GPT2Model,
    checkpoint,
    head_configs=head_configs,
    freeze_base_model=True,
    device_map="auto",
    adaptive_loss=False,
)
```

But given the way you've implemented loss calculation across heads, I'm unsure how to set up `TrainingArguments` accordingly. Specifically, what should the arguments be for `metric_for_best_model` and `compute_metrics` so that `Trainer` can properly index the loss information from the headed model during an evaluation step?

```py
args = TrainingArguments(
    ...,
    do_eval=True,
    metric_for_best_model="",
    greater_is_better=False,
    ...,
)
trainer = Trainer(
    args=args,
    ...,
    callbacks=[EarlyStoppingCallback(5)],
    compute_metrics=None
    ...,
)
```

Thanks again!

**Comment by yannikkellerde** [comment]
I never did early stopping with huggingface, so not an expert here (and this was not a concern when implementing HeadedModel). The HeadedModel still returns an [averaged loss](https://github.com/center-for-humans-and-machines/transformer-heads/blob/68a6f57845df8a6cab4bbeea93f1c802dfd28e52/transformer_heads/output.py#L34) over all the heads (which is the one that is trained on). So if you want to early stop based on this, this should work the same way as in standard transformer models. If you want to stop based on a specific head loss, I am not sure if this is easily possible. You might have to subclass Trainer for this.

**Comment by t-shoemaker** [comment]
Thanks for the response, all this makes sense. I haven't had a chance to tinker with this in full yet, but I think the solution will involve a custom `Trainer` regardless of whether one uses the averaged loss or some other mix of individual head losses. As things currently stand, `EarlyStoppingCallback` will [warn](https://github.com/huggingface/transformers/blob/98e8062df3cf82b367e8a243963e4afb0d9d3407/src/transformers/trainer_callback.py#L718) that it can't find a `*_loss` key. This doesn't break the training loop but it means that no evaluations are calculated. I suspect this is because there is no option to return a dictionary in the `HeadModel` forward pass, though if that's the case, it's unclear to me why the training dataset makes it through `Trainer` just fine.

I'll have to sleuth around in the `Trainer` file of the `transformers` code base and see what should be modified there in a subclass. In the meantime, I can close this. I appreciate it!
